src/components/fields/ArrayField.js

```diff
@@ -228,35 +284,69 @@ class ArrayField extends Component {
         event.preventDefault();
         event.target.blur();
       }
-      const {items} = this.state;
-      this.asyncSetState({
-        items: items.map((item, i) => {
+      const {formData: {items}, anyOfItems} = this.state;
+
+      const reorder = (items, newIndex) =>
+        items.map((item, i) => {
           if (i === newIndex) {
             return items[index];
           } else if (i === index) {
             return items[newIndex];
           } else {
             return item;
           }
-        })
-      }, {validate: true});
+        });
+
+      const newItems = {
+        items: reorder(items, newIndex)
+      };
+      const newAnyOfItems = reorder(anyOfItems, newIndex);
+
+      const newState = Object.assign({}, this.state,
+        {formData: newItems}, {anyOfItems: newAnyOfItems});
+      this.asyncSetState(newState, {validate: true});
     };
   };
 
   onChangeForIndex = (index) => {
     return (value) => {
-      this.asyncSetState({
-        items: this.state.items.map((item, i) => {
+      const items = {
+        items: this.state.formData.items.map((item, i) => {
           return index === i ? value : item;
         })
-      });
+      };
+      const newState = Object.assign({}, this.state, {formData: items});
+      this.asyncSetState(newState);
     };
   };
 
   onSelectChange = (value) => {
-    this.asyncSetState({items: value});
+    const newState = Object.assign({}, this.state, {formData: {items: value}});
+    this.asyncSetState(newState);
   };
 
+  anyOfOptions(anyOfItems) {
+    return anyOfItems.map(item => ({value: item.type, label: item.type}));
```

Comment:
Good spot, thanks! I think it should, yes. Maybe `item.title || item.type`, in case it's not defined...